src/OpenLoco/src/Map/TileManager.cpp

```diff
@@ -480,37 +490,519 @@ namespace OpenLoco::World::TileManager
+                        return Sub462B4FResult::noCollision;
+                    }
+                    if (_F0015C == reinterpret_cast<TileElement*>(-1))
+                    {
+                        // This is a no collision as it has removed all the collisions
+                        return Sub462B4FResult::allCollisionsRemoved;
+                    }
+                    // reset tile iterator to _F0015C (its saying we have removed a tile so pointer is stale)
+                    return Sub462B4FResult::collisionRemoved;
+                }
+            }
+            // some stuff getCollideDetails()
+            return Sub462B4FResult::collision;
+        };
+
+        auto checkSurfaceElement = [&clearFunc, &sub_462B4F, clearZ, baseZ, qt](const TileElement& el, const SurfaceElement& elSurface) -> Sub462B4FResult {
+            if (elSurface.isFlag5())
+            {
+                const TileElement* el2 = &el;
+                if (sub_462BB3(el2) != ClearResult::noCollision)
+                {
+                    return el2 == reinterpret_cast<TileElement*>(-1) ? Sub462B4FResult::collisionErrorSet : Sub462B4FResult::collision;
+                }
+            }
+            const auto waterZ = elSurface.water() * kMicroToSmallZStep;
+            if (waterZ != 0)
+            {
+                if (clearZ > elSurface.clearZ() && baseZ < waterZ + 4)
+                {
+                    _F00166 = _F00166 | (1 << 3);
+                    if (baseZ < waterZ)
+                    {
+                        _F00166 = _F00166 | (1 << 2);
+                        if (clearZ <= waterZ)
+                        {
+                            if (clearFunc.has_value())
+                            {
+                                _F0015C = nullptr;
+                                if (auto res = (clearFunc.value())(el); res != ClearResult::noCollision)
+                                {
+                                    if (res == ClearResult::collsionNoMessage)
+                                    {
+                                        GameCommands::setErrorText(StringIds::cannot_build_partly_above_below_water);
+                                    }
+                                    return Sub462B4FResult::collisionErrorSet;
+                                }
+                            }
+                            else
+                            {
+                                GameCommands::setErrorText(StringIds::cannot_build_partly_above_below_water);
+                                return Sub462B4FResult::collisionErrorSet;
+                            }
+                        }
+                    }
+                }
+            }
+            if (qt.getZQuarterOccupied() == 0xF)
+            {
+                return Sub462B4FResult::noCollision;
+            }
+
+            if (clearZ <= elSurface.baseZ())
+            {
+                _F00166 = _F00166 | (1u << 2);  // ELEMENT_IS_UNDERGROUND
```

Comment:
Check if this global is used in clear functions might need to be reset on restart of loop

Reply:
It's not